daemon/endpoint.go

```diff
@@ -1024,3 +1001,82 @@ func (d *Daemon) RemoveRestoredDNSRules(epID uint16) {
 }
+
+// endpointManager wraps the main endpointManager to allow endpoint deletion
+// to be triggered from other packages (like pkg/k8s) without having to
+// propagate various other subsystem objects around such as IPAM and the
+// monitor. In future if we either rearrange initialization logic or move
+// endpoint delete logic into pkg/endpointmanager, this type should go away.
+type endpointManager struct {
+	*endpointmanager.EndpointManager
+	d *Daemon
+}
+
+// NewEndpointManager wraps pkg/endpointmanager/EndpointManager to provide a
+// DeleteEndpoint() implementation that doesn't require the caller to pass
+// the IPAM implementation in.
+//
+// This works around the initialization order in the daemon of:
+// * First, EndpointManager
+// * After, K8sWatcher
+// * Later, IPAM.
+func NewEndpointManager(d *Daemon, epSync *endpointsynchronizer.EndpointSynchronizer) *endpointManager {
+	self := &endpointManager{d: d}
+	mgr := endpointmanager.NewEndpointManager(epSync)
+	epGC := option.Config.EndpointGCInterval
+	if epGC > 0 {
+		mgr = mgr.WithPeriodicEndpointGC(d.ctx, self, epGC)
+	}
+	self.EndpointManager = mgr
+	return self
+}
+
+func (e *endpointManager) DeleteEndpoint(ep *endpoint.Endpoint) int {
+	// Cancel any ongoing endpoint creation
+	e.d.endpointCreations.CancelCreateRequest(ep)
+
+	scopedLog := log.WithField(logfields.EndpointID, ep.ID)
+	errs := e.deleteEndpointQuiet(ep, endpoint.DeleteConfig{
+		// If the IP is managed by an external IPAM, it does not need to be released
+		NoIPRelease: ep.DatapathConfiguration.ExternalIpam,
+	})
+	for _, err := range errs {
+		scopedLog.WithError(err).Warn("Ignoring error while deleting endpoint")
+	}
+	return len(errs)
+}
+
+// deleteEndpointQuiet sets the endpoint into disconnecting state and removes
+// it from Cilium, releasing all resources associated with it such as its
+// visibility in the endpointmanager, its BPF programs and maps, (optional) IP,
+// L7 policy configuration, directories and controllers.
+//
+// Specific users such as the cilium-health EP may choose not to release the IP
+// when deleting the endpoint. Most users should pass true for releaseIP.
+func (e *endpointManager) deleteEndpointQuiet(ep *endpoint.Endpoint, conf endpoint.DeleteConfig) []error {
+	return ep.Delete(e.d, e.d.ipam, e, conf)
+}
+
+// Check verifies that endpoints are alive and healthy.
+func (e *endpointManager) Check(ep *endpoint.Endpoint) error {
+	// This only lives in endpointManager for convenience, the real
+	// intended home would be likely be pkg/datapath somewhere that
+	// understands how to gauge endpoint health from a node perspective.
+
+	// Be extra careful, we're only looking for one specific type of error
+	// currently: That the link has gone missing. Ignore other error to
+	// ensure that the caller doesn't unintentionally tear down the
+	// Endpoint thinking that it no longer exists.
+	_, err := netlink.LinkByName(ep.InterfaceName())
+	if _, ok := err.(netlink.LinkNotFoundError); ok {
```

Comment:
This will break ipvlan as in the ipvlan case, the `ep.InterfaceName()` will represent a device inside the container netns, not on the host.